hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java

```diff
@@ -0,0 +1,124 @@
+		final long nodeFee = tinycentsToTinybars(nodeFeeTinycents, rate) * multiplier;
+		final long serviceFee = tinycentsToTinybars(serviceFeeTinycents, rate) * multiplier;
+
+		if (networkFee < 0 || nodeFee < 0 || serviceFee < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
+
+		return new FeeObject(nodeFee, networkFee, serviceFee);
+	}
+
+	long tinycentsToTinybars(long amount, ExchangeRate rate) {
+		final var product = amount * rate.getHbarEquiv();
+		if (product < 0) {
+			return FeeBuilder.getTinybarsFromTinyCents(rate, amount);
+		}
+		return product / rate.getCentEquiv();
+	}
+
+	private long networkFeeInTinycents(UsageAccumulator usage, FeeComponents networkPrices) {
+		final var nominal = safeAccumulateThree(networkPrices.getConstant(),
+				usage.getUniversalBpt() * networkPrices.getBpt(),
+				usage.getNetworkVpt() * networkPrices.getVpt(),
+				usage.getNetworkRbh() * networkPrices.getRbh());
+		return ESTIMATOR_UTILS.nonDegenerateDiv(nominal, FEE_DIVISOR_FACTOR);
+	}
+
+	private long nodeFeeInTinycents(UsageAccumulator usage, FeeComponents nodePrices) {
+		final var nominal = safeAccumulateFour(nodePrices.getConstant(),
+				usage.getUniversalBpt() * nodePrices.getBpt(),
+				usage.getNodeBpr() * nodePrices.getBpr(),
+				usage.getNodeSbpr() * nodePrices.getSbpr(),
+				usage.getNodeVpt() * nodePrices.getVpt());
+		return ESTIMATOR_UTILS.nonDegenerateDiv(nominal, FEE_DIVISOR_FACTOR);
+	}
+
+	private long serviceFeeInTinycents(UsageAccumulator usage, FeeComponents servicePrices) {
+		final var nominal = safeAccumulateTwo(servicePrices.getConstant(),
+				usage.getServiceRbh() * servicePrices.getRbh(),
+				usage.getServiceSbh() * servicePrices.getSbh());
+		return ESTIMATOR_UTILS.nonDegenerateDiv(nominal, FEE_DIVISOR_FACTOR);
+	}
+
+	/* These verbose accumulators signatures are to avoid any performance hit from varargs */
+	long safeAccumulateFour(long base, long a, long b, long c, long d) {
+		if (base < 0 || a < 0 || b < 0 || c < 0 || d < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
+		var sum = safeAccumulateThree(base, a, b, c);
+		sum += d;
+		if (sum < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
+		return sum;
+	}
+
+	long safeAccumulateThree(long base, long a, long b, long c) {
+		if (base < 0 || a < 0 || b < 0 || c < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
+		var sum = safeAccumulateTwo(base, a, b);
+		sum += c;
+		if (sum < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
+		return sum;
+	}
+
+	long safeAccumulateTwo(long base, long a, long b) {
+		if (base < 0 || a < 0 || b < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
+		base += a;
+		if (base < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
+		base += b;
+		if (base < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
```

Comment:
nit: though unlikely, but If any of these cases happen, there must be some overflow. Will it be better to use a different message to indicate this case?

Reply:
True! Renamed to `OVERFLOW_ERROR` and changed message to,
```
A fee calculation step overflowed; the operation cannot be priced, and therefore cannot be performed
```